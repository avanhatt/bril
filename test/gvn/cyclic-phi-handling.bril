// Ported from llvm/test/Transforms/NewGVN/cyclic-phi-handling.ll

void main(arg : int, arg1 : int) {
bb:
  jmp bb3;

// While non-standard, llvm allows mutually dependent phi nodes
// Ensure we do not infinite loop trying to process them
bb3:
  tmp : int = phi arg1 tmp4 bb bb7;
  tmp4 : int = phi arg tmp bb bb7;
  tmp5 : int = call foo(tmp4, tmp);
  zero : int = const 0;
  tmp6 : int = eq tmp5 zero;
  tmp7 : bool =  not tmp6;
  br tmp7 bb7 bb8;

bb7:
  jmp bb3;

bb8:
  ret;
}

int foo(x : int, y : int) {
    z : int = add x y;
    ret z;
}