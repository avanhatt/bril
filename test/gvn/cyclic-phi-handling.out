void main(arg : int, arg1 : int) {
bb:
  jmp copy1;
bb3:
  tmp51: int = call foo(tmp41, tmp1);
  zero1: int = const 0;
  tmp61: int = eq tmp51 zero1;
  tmp71: bool = not tmp61;
  br tmp71 bb7 bb8;
bb7:
  jmp copy2;
bb8:
  ret ;
copy1:
  tmp1: int = id arg1;
  tmp41: int = id arg;
  jmp bb3;
copy2:
  tmp1: int = id tmp41;
  tmp41: int = id tmp1;
  jmp bb3;
}
int foo(x : int, y : int) {
b1:
  ret z1;
}
